crates/deltalake-core/src/operations/transaction/protocol.rs

```diff
@@ -0,0 +1,396 @@
+            }
+        };
+        Ok(())
+    }
+
+    /// Check if delta-rs can write to the given delta table.
+    pub fn can_write_to(&self, snapshot: &DeltaTableState) -> Result<(), TransactionError> {
+        // NOTE: writers must always support all required reader features
+        self.can_read_from(snapshot)?;
+
+        let required_features = match snapshot.min_writer_version() {
+            0 | 1 => None,
+            2 => Some(WRITER_V2.clone()),
+            3 => Some(WRITER_V3.clone()),
+            4 => Some(WRITER_V4.clone()),
+            5 => Some(WRITER_V5.clone()),
+            6 => Some(WRITER_V6.clone()),
+            _ => snapshot.writer_features().cloned(),
+        };
+
+        if let Some(features) = required_features {
+            let diff: Vec<_> = features
+                .difference(&self.writer_features)
+                .cloned()
+                .collect();
+            if !diff.is_empty() {
+                return Err(TransactionError::UnsupportedWriterFeatures(diff));
+            }
+        };
+        Ok(())
+    }
+
+    pub fn can_commit(
+        &self,
+        snapshot: &DeltaTableState,
+        actions: &[Action],
+    ) -> Result<(), TransactionError> {
+        self.can_write_to(snapshot)?;
+
+        // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#append-only-tables
+        let append_only_enabled = if snapshot.min_writer_version() < 2 {
+            false
+        } else if snapshot.min_writer_version() < 7 {
+            snapshot.table_config().append_only()
+        } else {
+            snapshot
+                .writer_features()
+                .ok_or(TransactionError::WriterFeaturesRequired)?
+                .contains(&WriterFeatures::AppendOnly)
+                && snapshot.table_config().append_only()
+        };
+        if append_only_enabled {
+            actions.iter().try_for_each(|action| match action {
+                Action::Remove(remove) if remove.data_change => {
+                    Err(TransactionError::DeltaTableAppendOnly)
+                }
+                _ => Ok(()),
+            })?;
+        }
+
+        Ok(())
+    }
+}
+
+/// The global protocol checker instance to validate table versions and features.
+pub static INSTANCE: Lazy<ProtocolChecker> = Lazy::new(|| {
+    let reader_features = HashSet::new();
+    // reader_features.insert(ReaderFeatures::ColumnMapping);
+
+    let mut writer_features = HashSet::new();
+    writer_features.insert(WriterFeatures::AppendOnly);
+    writer_features.insert(WriterFeatures::Invariants);
+    // writer_features.insert(WriterFeatures::CheckConstraints);
+    // writer_features.insert(WriterFeatures::ChangeDataFeed);
+    // writer_features.insert(WriterFeatures::GeneratedColumns);
+    // writer_features.insert(WriterFeatures::ColumnMapping);
+    // writer_features.insert(WriterFeatures::IdentityColumns);
+
+    ProtocolChecker::new(reader_features, writer_features)
+});
```

Comment:
Overall I like the API of `ProtocolChecker` and I think it's a great addition to the library. The motivation for this global I am struggling with however. I thought it might be to avoid allocations, but as far as I can tell, the common paths through `can_read` and `can_write_to` will clone a number of the related HashSets anyways, which has me perplexed.

I think I'm just not understanding, would you mind putting some more commentary above `INSTANCE` on the motivation for it so that I may be enlightened :smile:

Reply:
Happy to add some comments, just really quick here as well.

The cloning bothered me as well, and I'll see that I can fix that. The main motivation though is that with table features protocol support is essentially a negotiation between the client and the table. Since the table features we support lock in at compile time, there really is exactly one correct way the instance should look like, so i felt its less error prone if we just create it globally and track support for features in one place.

Reply:
added some comments and avoided cloning the sets.